Tweak Event.setCanceled To Match Convention

RlonRyan opened the pull request:
It is generalized Java convention to throw a UnsupportedOperationException whenever an operation is not supported, as in the case of Event.setCanceled() being called on an event where Event.isCancelable() returns false. The prior implementation threw an IllegalArgumentException, which did not make sense, given that it was not the argument that was the issue, but the fact that the method was even called (given that the method, as per its current implementation ignores the value of its parameter when determining if the exception should be thrown). Also, in the hopes of improving logs, the exception message has been changed as to include the canonical name of the Event subclass instance that was not cancelable. Thus, these changes should aid modders in determining the exact path (or mod interaction) that caused the given event instance as to be spawned and then subsequently attempted as to be canceled.

RlonRyan committed:
Tweak Event.setCanceled To Match Convention

It is generalized Java convention to throw a UnsupportedOperationException whenever an operation is not supported, as in the case of Event.setCanceled() being called on an event where Event.isCancelable() returns false. The prior implementation threw an IllegalArgumentException, which did not make sense, given that it was not the argument that was the issue, but the fact that the method was even called (given that the method, as per its current implementation ignores the value of its parameter when determining if the exception should be thrown). Also, in the hopes of improving logs, the exception message has been changed as to include the canonical name of the Event subclass instance that was not cancelable. Thus, these changes should aid modders in determining the exact path (or mod interaction) that caused the given event instance as to be spawned and then subsequently attempted as to be canceled.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=3926) <br/>All committers have signed the CLA.

RlonRyan commented:
[Example Gist](https://gist.github.com/RlonRyan/ea5418915e5de9f4482ed9ec23fc7384)

mezz commented on src/main/java/net/minecraftforge/fml/common/eventhandler/Event.java:
```diff
@@ -90,7 +91,10 @@ public void setCanceled(boolean cancel)
         {
-            throw new IllegalArgumentException("Attempted to cancel a non cancellable event");
+            throw new UnsupportedOperationException(
+                "Attempted to modify the unmodifiable canceled state of a non-cancelable event of type: "
```
I think the original wording is much more clear. Just add the class name to the end.

RlonRyan replied:
I was trying to capture the fact that the error would be thrown even if you pass `true` to the method. I concede, however, that my wording is not optimal.

Would say, `"Attempted to call Event#setCanceled() on a non-cancelable event of type: <type goes here>"` be better?

mezz replied:
That sounds good to me.

mezz commented on src/main/java/net/minecraftforge/fml/common/eventhandler/Event.java:
```diff
@@ -78,9 +78,10 @@ public boolean isCanceled()
-    /**
-     * Sets the state of this event, not all events are cancelable, and any attempt to
-     * cancel a event that can't be will result in a IllegalArgumentException.
+    /*
```
Javadocs should start with `/**`

RlonRyan replied:
Oh dear me, how could I have missed that...

mezz reviewed (changes requested):
Makes sense, needs some minor changes.

RlonRyan committed:
Improve Exception Message Wording

This commit clarifies the wording of the message associated with the UnsupportedOperationException thrown by the Event#setCanceled() method when the given event instance is non-cancelable. Furthermore, this commit also corrects the missing second star from the JavaDoc for the aforementioned method that had turned the JavaDoc into a block comment instead. Thank you @mezz for your assistance with this patch.

mezz commented:
Looks good to me, thanks.

RlonRyan commented:
Oh, it looks like the [documentation](http://mcforge.readthedocs.io/en/latest/events/intro/#canceling-results) will have to be updated after this...

mezz commented:
@RlonRyan could you submit a PR to fix the docs here? 
https://github.com/MinecraftForge/Documentation/blob/master/docs/events/intro.md

RlonRyan commented:
Sorry about the delay Mezz, I've been otherwise occupied recently. I'll look into making that PR for you now though.

mezz commented:
No rush. Thanks.
